From e10370efdce8698e87b729027b35bd93e1b68498 Mon Sep 17 00:00:00 2001
From: Tobias Waldekranz <tobias@waldekranz.com>
Date: Wed, 24 Apr 2024 21:35:26 +0200
Subject: [PATCH 26/33] net: dsa: mv88e6xxx: Fix port policy config on 6393X
Organization: Addiva Elektronik

mv88e6393x_port_policy_{read,write} expect the `pointer` argument to
be "pre-shifted" 8 bits.

Therefore, make sure mv88e6393x_port_set_policy adheres to this
requirement.

Before this change, no policy actions from offset 1 could be
configured. In particular, this meant that VTU policy trapping was
never enabled on DSA ports, which broke standalone port isolation in
multichip switch trees made up of 6393X decices.

Fixes: 6584b26020fc ("net: dsa: mv88e6xxx: implement .port_set_policy for Amethyst")
---
 drivers/net/dsa/mv88e6xxx/port.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dsa/mv88e6xxx/port.c b/drivers/net/dsa/mv88e6xxx/port.c
index aba78838171b..3f5759992baf 100644
--- a/drivers/net/dsa/mv88e6xxx/port.c
+++ b/drivers/net/dsa/mv88e6xxx/port.c
@@ -1793,12 +1793,12 @@ int mv88e6393x_port_set_policy(struct mv88e6xxx_chip *chip, int port,
 	shift %= 8;
 	mask >>= ptr * 8;
 
-	err = mv88e6393x_port_policy_read(chip, port, ptr, &reg);
+	err = mv88e6393x_port_policy_read(chip, port, ptr << 8, &reg);
 	if (err)
 		return err;
 
 	reg &= ~mask;
 	reg |= (val << shift) & mask;
 
-	return mv88e6393x_port_policy_write(chip, port, ptr, reg);
+	return mv88e6393x_port_policy_write(chip, port, ptr << 8, reg);
 }
-- 
2.43.0

